### PR TITLE
Export video transcripts

### PR DESCRIPTION
**Pull request by mushtaqak**
Related platform [PR#17488](https://github.com/edx/edx-platform/pull/17488)

[EDUCATOR-1789](https://openedx.atlassian.net/browse/EDUCATOR-1789)

Sandbox URL: https://studio-exporttranscripts.sandbox.edx.org/export/course-v1:edX+DemoX+Demo_Course

### CODE REVIEW

**Comment by mushtaqak** [comment]
@muhammad-ammar @Qubad786 This is ready for review.

File: edxval/api.py [hunk 1]
```diff
@@ -831,37 +823,70 @@ def export_to_xml(video_ids, course_id=None, external=False):
...
     # create transcripts node only when we have transcripts for a video
     if video_transcripts.exists():
         transcripts_el = SubElement(video_el, 'transcripts')
 
     exported_language_codes = []
     for video_transcript in video_transcripts:
         if video_transcript.language_code not in exported_language_codes:
+            edx_video_id = video_transcript.video.edx_video_id
+            language_code = video_transcript.language_code
+            transcript_url = video_transcript.transcript.url
```
Comment (Qubad786) [inline edxval/api.py]: Why this?

File: edxval/api.py [hunk 2]
```diff
@@ -831,37 +823,70 @@ def export_to_xml(video_ids, course_id=None, external=False):
...
+            file_format=file_format
+        )
+        transcript_data = get_video_transcript_data(edx_video_id, language_code)
+        if transcript_data:
+            transcript_content = transcript_data['content']
+            with resource_fs.open(transcript_name, 'wb') as f:
+                f.write(transcript_content)
+
+
+def create_transcripts_xml(video_id, video_el, resource_fs=None, transcript_dir=None):
```
Comment (Qubad786) [inline edxval/api.py]: Is there any reason for making ` resource_fs=None, transcript_dir=None` optionals?

File: edxval/api.py [hunk 3]
```diff
@@ -831,37 +823,70 @@ def export_to_xml(video_ids, course_id=None, external=False):
...
+
+            create_trancript_file(edx_video_id, language_code, file_format, transcript_dir, resource_fs)
+
             SubElement(
                 transcripts_el,
                 'transcript',
                 {
-                    'video_id': video_transcript.video.edx_video_id,
+                    'video_id': edx_video_id,
                     'file_name': video_transcript.transcript.name,
```
Comment (Qubad786) [inline edxval/api.py]: we don't need this.

Comment (Qubad786) [reply edxval/api.py]: we can only send language_code, file_format, and  provider (video id can be retrieved from the component)

File: test-requirements.txt [hunk 4]
```diff
@@ -1,5 +1,7 @@
 coverage==3.7.1
 ddt==0.8.0
 django-nose==1.4.4
+fs==2.0.17
 mock==1.0.1
+path.py==8.2.1
```
Comment (Qubad786) [inline test-requirements.txt]: can we use `os.path`?

File: edxval/tests/test_api.py [hunk 5]
```diff
@@ -2030,6 +2038,54 @@ def test_delete_video_transcript(self, mock_logger):
...
+        self.assertTrue(u'super-soaker-en.srt' in file_system.listdir(STATIC_PATH))
+
+    @data(
+        ('invalid-video-id', 'invalid-language-code', None, False),
+        ('super-soaker', 'invalid-language-code', None, False),
+        ('super-soaker', 'en', None, True),
+        ('super-soaker', 'en', STATIC_DIR, False)
+    )
+    @unpack
+    def test_no_create_transcript_file(self, video_id, language_code, transcript_dir, is_resource_fs):
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: If missing one of the `transcript_dir` or `is_resource_fs`, is an invalid scenario then, we should not make them optional in the first place and this test wont also be needed.

File: edxval/tests/test_api.py [hunk 6]
```diff
@@ -2030,6 +2038,54 @@ def test_delete_video_transcript(self, mock_logger):
...
+        # Create transcript file now.
+        api.create_trancript_file(
+            edx_video_id=constants.VIDEO_DICT_FISH['edx_video_id'],
+            language_code='en',
+            file_format=TranscriptFormat.SRT,
+            transcript_dir=STATIC_DIR,
+            resource_fs=file_system
+        )
+
+        # Verify transcript is file created.
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: typo: Verify the transcript file.

File: edxval/tests/test_api.py [hunk 7]
```diff
@@ -2030,6 +2038,54 @@ def test_delete_video_transcript(self, mock_logger):
...
+        self.assertTrue(u'super-soaker-en.srt' in file_system.listdir(STATIC_PATH))
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: lets also assert the content by opening the created file.

File: edxval/tests/test_api.py [hunk 8]
```diff
@@ -983,44 +1016,16 @@ def test_basic(self, course_id, image):
...
-            api.export_to_xml([constants.VIDEO_DICT_FISH['edx_video_id']], course_id),
+            api.export_to_xml(constants.VIDEO_DICT_FISH['edx_video_id'], course_id),
             expected
         )
 
     def test_unknown_video(self):
-        with self.assertRaises(ValVideoNotFoundError):
-            api.export_to_xml(["unknown_video"])
-
-    def test_with_multiple_video_ids(self):
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: nice!

File: edxval/tests/test_api.py [hunk 9]
```diff
@@ -901,7 +908,9 @@ def test_existing_video_in_destination_course_id(self):
 
 @ddt
 class ExportTest(TestCase):
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: We are missing a test case that asserts the exported transcript XML along with the content in the file system.

File: edxval/api.py [hunk 10]
```diff
@@ -831,37 +823,70 @@ def export_to_xml(video_ids, course_id=None, external=False):
...
+    """
+    Writes transcript file to file system.
 
-def create_transcripts_xml(video_ids, video_el):
+    Arguments:
+        edx_video_id (str): Video id of the video transcript file is attached.
+        language_code (str): Language code of the transcript.
+        file_format (str): File format of the transcript file.
+        transcript_dir (str): The Directory to store transcript file.
+        resource_fs (SubFS): The file system to store transcripts.
```
[the remark this answers is not in the document]
Comment (mushtaqak) [reply edxval/api.py]: `SubFS` is used which is child class of `FSOS`

File: edxval/tests/test_api.py (hunk 9 above)
Comment (mushtaqak) [reply edxval/tests/test_api.py]: Added

File: edxval/tests/test_api.py [hunk 11]
```diff
@@ -951,76 +966,108 @@ def get_child_tags(elem):
...
-        video_ids = ['super-soaker', 'external_video_id']
-        expected = self.parse_xml("""
-            <video_asset client_video_id="Shallow Swordfish" duration="122.0" image="">
-                <encoded_video bitrate="22" file_size="11" profile="mobile" url="http://www.meowmix.com" />
-                <encoded_video bitrate="44" file_size="33" profile="desktop" url="http://www.meowmagic.com" />
-                <encoded_video bitrate="0" file_size="100" profile="hls" url="https://www.tmnt.com/tmnt101.m3u8" />
+        language_code = 'en'
+        video_id = constants.VIDEO_DICT_FISH['edx_video_id']
+        transcript_files = {'de': u'super-soaker-de.sjson', 'en': u'super-soaker-en.srt'}
+        expected_transcript_path = self.temp_dir + STATIC_PATH + '/{transcript_file_name}'
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: Use either str.format OR `os.path.join`

Comment (mushtaqak) [reply edxval/tests/test_api.py]: `self.temp_dir ` is an absolute path, so `os.path.join(self.temp_dir, STATIC_PATH )` would only yield `/static'. 

Using the `fs.path.combine` which is similar method to `os.path.join` for fs.

**Pull request merged**